src/django_components/provide.py

```diff
@@ -87,8 +90,79 @@ def render(self, context: Context, name: str, **kwargs: Any) -> SafeString:
         # add the provided kwargs into the Context.
         with context.update({}):
             # "Provide" the data to child nodes
-            set_provided_context_var(context, name, kwargs)
+            provide_id = set_provided_context_var(context, name, kwargs)
 
-            output = self.nodelist.render(context)
+            # `managed_provide_cache` will remove the cache entry at the end if no components reference it.
+            with managed_provide_cache(provide_id):
+                output = self.nodelist.render(context)
 
         return output
+
+
+def get_injected_context_var(
```

Comment:
`get_injected_context_var`  and `set_provided_context_var` were moved here from `context.py`.